File: Linkpoint/src/main/java/com/linkpoint/render/glres/GLLoadQueue.java
package com.linkpoint.render.glres;

import com.linkpoint.render.RenderContext;
import com.linkpoint.res.collections.WeakQueue;
import javax.annotation.Nonnull;

public abstract class GLLoadQueue {
    final WeakQueue<GLLoadable> loadQueue = new WeakQueue<>();

    interface GLLoadHandler {
        void GLResourceLoaded(GLLoadable gLLoadable);
    }

    interface GLLoadable {
        void GLCompleteLoad();

        int GLGetLoadSize();

        int GLLoad(RenderContext renderContext, GLLoadHandler gLLoadHandler);
    }

    public abstract void RunLoadQueue(@Nonnull RenderContext renderContext);

    public void StopLoadQueue() {
        this.loadQueue.clear();
    }

    public void add(@Nonnull GLLoadable gLLoadable) {
        this.loadQueue.offer(gLLoadable);
    }

    public void remove(@Nonnull GLLoadable gLLoadable) {
        this.loadQueue.remove(gLLoadable);
    }
}

File: Linkpoint/src/main/kotlin/com/linkpoint/render/glres/GLLoadQueue.kt
package com.linkpoint.render.glres

import com.linkpoint.render.RenderContext
import com.linkpoint.res.collections.WeakQueue
import javax.annotation.Nonnull

abstract class GLLoadQueue {
    final WeakQueue<GLLoadable> loadQueue = WeakQueue<>()

    interface GLLoadHandler {
        fun GLResourceLoaded(gLLoadable: GLLoadable)
    }

    interface GLLoadable {
        fun GLCompleteLoad()

        fun GLGetLoadSize(): Int

        fun GLLoad(renderContext: RenderContext, gLLoadHandler: GLLoadHandler): Int
    }

    public abstract Unit RunLoadQueue(RenderContext renderContext)

    fun StopLoadQueue() {
        this.loadQueue.clear()
    }

    fun add(gLLoadable: GLLoadable) {
        this.loadQueue.offer(gLLoadable)
    }

    fun remove(gLLoadable: GLLoadable) {
        this.loadQueue.remove(gLLoadable)
    }
}
